edgepi-pwm.js

```diff
@@ -0,0 +1,110 @@
+module.exports = function (RED) {
+  const rpc = require("@edgepi-cloud/edgepi-rpc");
+
+  function PWMNode(config) {
+    RED.nodes.createNode(this, config);
+    const node = this;
+    const pwmPin = rpc.PWMPins[config.pwmPin];
+
+    setInitialConfigs(config, node)
+      .then((pwm) => {
+        node.on("input", async function (msg, send, done) {
+          node.status({ fill: "green", shape: "dot", text: "input received" });
+          try {
+            await updateConfigsOnInput(msg, pwm, config);
+            msg = {
+              payload: await pwm.getDutyCycle(pwmPin),
+              pwmPin: pwmPin,
+              frequency: await pwm.getFrequency(pwmPin),
+              polarity: await pwm.getPolarity(pwmPin),
+              enabled: await pwm.getEnabled(pwmPin),
+            };
+          } catch (error) {
+            console.error(error);
+            msg.payload = error;
+          }
+          send(msg);
+          if (done) done();
+        });
+
+        node.on("close", async () => {
+          try {
+            await pwm.close(pwmPin);
+          } catch (error) {
+            console.error(error);
+          }
+        });
+      })
+      .catch((error) => {
+        console.error(error);
+        node.status({
+          fill: "red",
+          shape: "ring",
+          text: "Initialization error",
+        });
+      });
+
+    async function setInitialConfigs(config, node) {
+      const transport =
+        config.transport === "Network"
+          ? `tcp://${config.tcpAddress}:${config.tcpPort}`
+          : "ipc:///tmp/edgepi.pipe";
+      const configArg = {
+        pwmNum: config.pwmPin,
+        frequency: parseFloat(config.frequency),
+        dutyCycle: parseFloat(config.dutyCycle) / 100,
+        polarity: config.polarity,
+      };
+
+      try {
+        const pwm = new rpc.PWMService(transport);
+        await pwm.initPwm(pwmPin);
+        await pwm.setConfig(configArg);
+        if (config.state === "enabled") {
+          await pwm.enable(pwmPin);
+        } else {
+          await pwm.disable(pwmPin);
+        }
+        return pwm;
+      } catch (error) {
+        error.message =
+          "Error setting initial configurations: " + error.message;
+        throw error;
+      }
+    }
+
+    async function updateConfigsOnInput(msg, pwm, config) {
+      let inputConfig = { pwmNum: pwmPin };
+      const { frequency, payload, polarity, enabled } = msg;
+
+      if (frequency) {
+        inputConfig.frequency = parseFloat(frequency);
+      }
+      if (payload) {
+        inputConfig.dutyCycle = parseFloat(payload) / 100;
+      }
+      if (polarity) {
+        inputConfig.polarity = rpc.Polarity[polarity];
+      }
+      if (enabled !== undefined) {
+        enabled ? await pwm.enable(pwmPin) : await pwm.disable(pwmPin);
+      }
```

Comment:
why for `enabled` we have a check for if it's not `undefined` but the others there is no check? what makes `enabled` different from the others in `msg`?

Reply:
The first 4 if statements are checking for truthiness, which includes `false` as well as undefined. Here, I'm doing it to check that the variables exists. `enabled` is a boolean variable. Whether it's true or false, I still want the code in the if block to execute. I could check all the variables for `undefined`, but since the rest are not boolean it's not necessary.